### PR TITLE
Support for generation of project tags file

### PR DESCRIPTION
**Pull request by tuna-f1sh**
Considering the number of project files spread in different locations when developing an Arduino project, proper use of tags can be difficult; resolving beyond local functions.

I've added automatic generation of a tags file, which includes:

* Standard source in project dir (.c, .cpp, .h)
* Arduino source in project dir (.ide, .pde)
* Arduino core based on detected project core from Arduino install.
* Included Arduino libraries from user library folder.

As a Vim user I find this hugely useful and think it would be a useful addition for others. Target has been added as `make tags`. There may be some better practices, for some things (I'm not sure about using `pwd` in the shells commands?). I've included a few variables, see updates variables file.

John.

### CODE REVIEW

File: Arduino.mk [hunk 1]
```diff
@@ -1562,6 +1578,21 @@ generate_assembly: $(OBJDIR)/$(TARGET).s
 generated_assembly: generate_assembly
 		@$(ECHO) "\"generated_assembly\" target is deprecated. Use \"generate_assembly\" target instead\n\n"
 
+.PHONY: tags
+tags:
+	rm -f $(shell pwd)/$(TAGS_FILE)
+	@$(ECHO) "Generating tags for local sources (IDO an PDE files as C++): "
```
Comment (sej7278) [inline Arduino.mk]: should this be:

I**N**O an**d** PDE files as C++

**Comment by tuna-f1sh** [comment]
Thanks good spot, now fixed.

File: Arduino.mk [hunk 2]
```diff
@@ -1562,6 +1578,21 @@ generate_assembly: $(OBJDIR)/$(TARGET).s
 generated_assembly: generate_assembly
 		@$(ECHO) "\"generated_assembly\" target is deprecated. Use \"generate_assembly\" target instead\n\n"
 
+.PHONY: tags
+tags:
+	rm -f $(shell pwd)/$(TAGS_FILE)
```
Comment (sej7278) [inline Arduino.mk]: i'm a bit worried about this accidentally deleting stuff if $(TAGS_FILE) is undefined, is it just a single file?

if so then ```rm -f $(TAGS_FILE)``` i guess is a bit safer.....?

also is this going to be in the same directory as the Makefile and .ino, or in $(OBJDIR)? i.e. will "make clean" remove it?

**Comment by sej7278** [comment]
@sudar is more of a vim guy, care to chime in?

File: Arduino.mk (hunk 2 above)
Comment (sudar) [reply Arduino.mk]: I agree. We need a guard condition to prevent accidental deletion.

**Comment by tuna-f1sh** [comment]
Thanks. I've taken this into account - I agree that even with the definition check of the tags file, it could lead to problems. I use your syntax plugin @sudar so I thought you might like it 👍 .

It's now `rm -f $(TAGS_FILE)` and only if the TAGS_FILE exists. One could get away without deleting the tags file completely as ctags replaces by default but I make use of the append flag to do a multi-line operation. It is common practice for the tags file to be in the root of the project directory, so `make clean` will not remove it - I think this is desirable however.

**Comment by sej7278** [comment]
odd, i'm sure merging #524 conflicts with this, but github isn't picking it up.

does it only pick up changes from master when a PR is updated?

**Comment by tuna-f1sh** [comment]
I've merged the upstream to check but there were no conflicts.

**Comment by sej7278** [comment]
looks good to me 👍

**Comment by sej7278** [comment]
can we merge this now @sudar ?

**Comment by sudar** [comment]
@sej7278 Just merged it.

**Pull request merged**